### PR TITLE
Issue 660

### LINKED ISSUE #660: Backend for follow up report

### PR DESCRIPTION
**Pull request by asekito**
Closes #660

This one is a little messy. I was going to refactor when there was more time but I figured we wanted this completed sooner than later.
Follow ups calculations and results mimic the same logic as interactions and the SQL was tweaked to allow a unique total number of clients upon search through followUps and clientInteractions table so no clients were double counted.

Let me know if you want better variable names. Will refactor when I can (if I should -- still have to look). Wanted this out and working.

Note keys surrounded with ** are the new ones for follow ups specifically. No change in the variables for interactions to avoid frontend conflict.

Sample output of end point response to client side:
```
{
  **followUpProgramTotals**: (9) [{…}, {…}, {…}, {…}, {…}, {…}....]
  **followUpServicesTotal**: (41) [{…}, {…}, {…}, {…}, {…}, {…}....]
  grandTotal: {
    **numClients**: 4
    numInteractions: 3
    totalDuration: "05:30:30" // for interactions
    **totalFollowUpDuration**: "07:00:00" // for follow ups
    **totalFollowUpSeconds**: 25200
    totalInteractionSeconds: 19830
    programs: (9) [{…}, {…}, {…}, {…}, {…}.....]
    reportParameters: {start: null, end: null}
  }
  services: (41) [{…}, {…}, {…}, {…}, {…},....]
}
```

---------------------------------------

Sample for followUpServicesTotal:

```
**followUpServicesTotal**: Array(41)
  0: {serviceName: "Chronic Disease Screenings", programName: "Preventive Health", numFollowUps: 2, numClients: 0, 
  totalFollowUpSeconds: 7200, …}
  1:
    numClients: 0
    **numFollowUps**: 2
    programId: 1
    programName: "Preventive Health"
    serviceId: 15
    serviceName: "Chronic Care Guidance"
    **totalDuration**: "02:00:00"
    **totalFollowUpSeconds**: 7200
  2: {serviceName: "Tobacco Prevention and Cessation", programName: "Preventive Health", numFollowUps: 0, numClients: 0, 
  totalFollowUpSeconds: 0, …}
.......
```

---------------------------------------

Sample for followUpProgramTotals:

```
  **followUpProgramTotals**: Array(9)
  0:
    numClients: 2
    **numFollowUps**: 4
    programId: 1
    programName: "Preventive Health"
    **totalDuration**: "04:00:00"
    **totalFollowUpSeconds**: 14400
  1: {programId: 2, programName: "Nutrition / CRYS", numFollowUps: 1, numClients: 1, totalFollowUpSeconds: 10800, …}
  2: {programId: 3, programName: "Immigration", numFollowUps: 0, numClients: 0, totalFollowUpSeconds: 0, …}
.......
```

### CODE REVIEW

**Review by joeldenning — COMMENTED** [review]
Looks like we might be missing the number of clients who've received follow ups per **service**? I see it for program, but not service

File: backend/apis/reports/interactions-by-service.api.js [hunk 1]
```diff
@@ -71,7 +71,7 @@ app.get(`/api/reports/interactions-by-service`, (req, res) => {
       ;
 
       -- num total clients
-      SELECT COUNT(DISTINCT clientInteractions.clientId) numClients
+      SELECT (COUNT(DISTINCT clientInteractions.clientId) + (SELECT COUNT(DISTINCT followUps.clientId) FROM followUps WHERE followUps.clientId <> clientInteractions.clientId)) numClients
```
[the remark this answers is not in the document]
Comment (asekito) [reply backend/apis/reports/interactions-by-service.api.js:74]: Oh I thought it was working earlier yesterday when I was testing it out but no it isn't. What if I use two selects and use a UNION operator?

File: backend/apis/reports/interactions-by-service.api.js [hunk 2]
```diff
@@ -28,7 +28,7 @@ app.get(`/api/reports/interactions-by-service`, (req, res) => {
         (
           SELECT COUNT(*) totalInteractions, serviceId
           FROM clientInteractions
-          WHERE isDeleted = false AND dateOfInteraction >= ? AND dateOfInteraction <= ?
+          WHERE isDeleted = false AND dateOfInteraction >= '2000-01-01' AND dateOfInteraction <= '2020-12-31'
```
[the remark this answers is not in the document]
Comment (asekito) [reply backend/apis/reports/interactions-by-service.api.js:31]: Yes. Will change back to the binding parameter.

File: backend/apis/reports/interactions-by-service.api.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (joeldenning) [reply backend/apis/reports/interactions-by-service.api.js:74]: Yeah UNION looks like it could be a good approach

**Review by joeldenning — APPROVED** [review]
This looks good to me. Anything left before I merge this?

**Comment by asekito** [comment]
Nope looks good to go except it says there was a CI error?

**Comment by joeldenning** [comment]
The CI error is related the yarn lockfile, but it's weird that it worked in the push build, but not the PR build. This doesn't seem to be your fault - I'll merge this now.

**Pull request merged**